### PR TITLE
Log console output during replay to file #985

### PR DESCRIPTION
**Pull request by cogutvalera**
PR for **"Log console output during replay to file #985 https://github.com/bitshares/bitshares-core/issues/985"**

### CODE REVIEW

**Comment by abitmore** [comment]
Can you please briefly explain why change like this? There is no comment in your code.

**Comment by cogutvalera** [comment]
I can add comments right into code if required. 
After debugging and researching code I've found the reason of this issue.
1. During `replay` we can see logs in console but not in file because of next reason: `fc::file_appender::impl::out` (`ofstream`) is opened after `rotate_files`. 
2. So this means that `ofstream` will be opened after `replay` is completed and finished. Thus after `replay` we can see the same logs in console and file.
3. Now after my changes `ofstream` is opened before `replay` happens. So this solution makes possible to see logs not only in console during `replay` but also we can see the same logs in file according to configuration in `config.ini` or `logging.ini`.

File: src/log/file_appender.cpp [hunk 1]
```diff
@@ -141,9 +160,11 @@ namespace fc {
       {
          fc::create_directories(my->cfg.filename.parent_path());
 
-         if(!my->cfg.rotate)
-            my->out.open( my->cfg.filename, std::ios_base::out | std::ios_base::app);
+         {
+            fc::scoped_lock<boost::mutex> lock( my->slock );
```
Comment (jmjatlanta) [inline src/log/file_appender.cpp]: As this is a constructor, I do not see the need for a lock here.

File: src/log/file_appender.cpp [hunk 2]
```diff
@@ -141,9 +160,11 @@ namespace fc {
       {
          fc::create_directories(my->cfg.filename.parent_path());
 
-         if(!my->cfg.rotate)
```
Comment (jmjatlanta) [inline src/log/file_appender.cpp]: I haven't walked all of the code, but if rotate == true, will removing this cause open() to be called twice (once here, and once in rotate_files()) ?

**Review by jmjatlanta — CHANGES_REQUESTED** [review]

File: src/log/file_appender.cpp (hunk 2 above)
Comment (cogutvalera) [reply src/log/file_appender.cpp]: `if rotate` check was removed, and twice won't be open because there is a check in `open_log_filename` method

File: src/log/file_appender.cpp (hunk 1 above)
Comment (cogutvalera) [reply src/log/file_appender.cpp]: we have different threads, that's why I've added lock here, IMHO `slock` mutex is supposed to be used for log files within different threads, maybe I'm wrong.

Thanks !

File: src/log/file_appender.cpp [hunk 3]
```diff
@@ -57,21 +69,30 @@ namespace fc {
             }
          }
 
+         void open_log_filename()
```
Comment (abitmore) [inline src/log/file_appender.cpp]: This function name is confusing. I'd recommend rename it to `open_log_file()` or something else that can describe the logic better.

In addition, the check for file name inside the function made it even harder to understand, I think it would be better to move out the check, e.g. have an `open()` simply opens a file or perhaps close then reopen, and only call it when necessary (aka with a check).

In short, try to simplify the code.

File: src/log/file_appender.cpp [hunk 4]
```diff
@@ -23,6 +23,10 @@ namespace fc {
       private:
          future<void>               _rotation_task;
          time_point_sec             _current_file_start_time;
+         fc::time_point             _now;
+         fc::time_point_sec         _start_time;
+         string                     _timestamp_string;
+         fc::path                   _old_log_filename, _log_filename, _link_filename;
```
Comment (abitmore) [inline src/log/file_appender.cpp]: For member variables, please declare only one variable per line.

File: src/log/file_appender.cpp [hunk 5]
```diff
@@ -159,7 +176,9 @@ namespace fc {
       std::stringstream line;
       //line << (m.get_context().get_timestamp().time_since_epoch().count() % (1000ll*1000ll*60ll*60))/1000 <<"ms ";
       line << string(m.get_context().get_timestamp()) << " ";
-      line << std::setw( 21 ) << (m.get_context().get_thread_name().substr(0,9) + string(":") + m.get_context().get_task_name()).c_str() << " ";
+      line << std::setw( 21 ) << 
+         (m.get_context().get_thread_name().substr(0,9) + string(":") + m.get_context().get_task_name()).c_str() 
+      << " ";
```
Comment (abitmore) [inline src/log/file_appender.cpp]: The code style here looks a bit ugly. I'd prefer
```
line << "something"
     << "something else"
     << "the end";       
```

File: src/log/file_appender.cpp [hunk 6]
```diff
@@ -141,9 +160,7 @@ namespace fc {
       {
          fc::create_directories(my->cfg.filename.parent_path());
```
Comment (abitmore) [inline src/log/file_appender.cpp]: Why not move the whole logic into constructor of `impl` class for simplicity? All operation is done on `my` anyway.

**Review by abitmore — DISMISSED** [review]
Need better abstraction / modularization.

File: src/log/file_appender.cpp (hunk 6 above)
Comment (cogutvalera) [reply src/log/file_appender.cpp]: I thought the same earlier, I agree it would be better ! Thanks !

**Comment by cogutvalera** [comment]
pushed new commit **"more improvements"**

Thanks !

**Comment by pmconrad** [comment]
> So this means that ofstream will be opened after replay is completed and finished.

Why is this so?
file_appender::impl calls rotate_files asynchronously from the constructor. Either there's something wrong with the call, or the async excution happens only after replay. In the latter case, the obvious (and much simpler) fix would be to make the initial rotate_files call synchronous.

File: src/log/file_appender.cpp [hunk 7]
```diff
@@ -31,19 +37,46 @@ namespace fc {
...
+            }
+            catch( ... )
+            {
+               std::cerr << "error opening log file: " << cfg.filename.preferred_string() << "\n";
+            }
 
+            if( cfg.rotate )
+            {
+               FC_ASSERT( cfg.rotation_interval >= seconds( 1 ) );
+               FC_ASSERT( cfg.rotation_limit >= cfg.rotation_interval );
```
Comment (abitmore) [inline src/log/file_appender.cpp]: I think these assertions should be put before creating log file, even before `setup_log_filenames`.

File: src/log/file_appender.cpp [hunk 8]
```diff
@@ -31,19 +37,46 @@ namespace fc {
...
 
+            try
+            {
+               fc::create_directories(cfg.filename.parent_path());
 
+               open_log_file();
+            }
+            catch( ... )
+            {
+               std::cerr << "error opening log file: " << cfg.filename.preferred_string() << "\n";
```
Comment (abitmore) [inline src/log/file_appender.cpp]: Is the appender usable after caught an exception here? What will happen when trying to write?

File: src/log/file_appender.cpp [hunk 9]
```diff
@@ -82,15 +111,18 @@ namespace fc {
...
                }
-               remove_all(link_filename);  // on windows, you can't delete the link while the underlying file is opened for writing
-               out.open( log_filename, std::ios_base::out | std::ios_base::app );
-               create_hard_link(log_filename, link_filename);
+               if (_old_log_filename != _log_filename)
+               {
+                  _old_log_filename = _log_filename;
+                  
+                  open_log_file();
+               }
```
Comment (abitmore) [inline src/log/file_appender.cpp]: The logic seems a bit strange.
* Is there a chance that the file got closed but not reopened?
* Is there a chance that the file be opened again when not closed?

**Comment by abitmore** [comment]
> the obvious (and much simpler) fix would be to make the initial rotate_files call synchronous.

Probably because `rotate_files()` also removes old files that could take some time, thus not good to call on initialization. However, even it's the case, we can skip the removal on first call.

By the way, I guess the `rotate_files(1)` call is defensive? Looks like it will never be called.

**Comment by cogutvalera** [comment]
(just for the info) I've pushed 2 new commits:

1. reverted all changes made before to current master branch
2. Log console output during replay to file #985 (2nd simpler solution)

File: src/log/file_appender.cpp [hunk 10]
```diff
@@ -137,18 +134,7 @@ namespace fc {
    file_appender::file_appender( const variant& args ) :
      my( new impl( args.as<config>( FC_MAX_LOG_OBJECT_DEPTH ) ) )
    {
-      try
-      {
-         fc::create_directories(my->cfg.filename.parent_path());
```
Comment (pmconrad) [inline src/log/file_appender.cpp:142]: I think this must be moved to impl::impl(), not simply removed.

File: src/log/file_appender.cpp [hunk 11]
```diff
@@ -137,18 +134,7 @@ namespace fc {
...
-
-         if(!my->cfg.rotate)
```
Comment (pmconrad) [inline src/log/file_appender.cpp:144]: ...and this shouldn't be removed.

**Review by pmconrad — CHANGES_REQUESTED** [review]
*If we decide to merge the second version*, you should ``rebase -i`` to get rid of the first version. But not yet.

**Comment by cogutvalera** [comment]
> If we decide to merge the second version, you should rebase -i to get rid of the first version. But not yet.

IMHO the 2nd solution is simpler and it works as required, then why to rebase "not yet" ?

File: src/log/file_appender.cpp [hunk 12]
```diff
@@ -34,16 +34,25 @@ namespace fc {
...
+            {
+               FC_ASSERT( cfg.rotation_interval >= seconds( 1 ) );
+               FC_ASSERT( cfg.rotation_limit >= cfg.rotation_interval );
 
+               rotate_files( true );
+            }
 
+            try
+            {
+               fc::create_directories(cfg.filename.parent_path());
```
Comment (pmconrad) [inline src/log/file_appender.cpp]: create_directories should happen before rotate_files.

**Comment by pmconrad** [comment]
The idea was to preserve the original version in case my ideas didn't work. I'm ok with the rebase if it does work as intended.

**Review by pmconrad — APPROVED** [review]

**Comment by pmconrad** [comment]
Thanks! @abitmore @jmjatlanta please re-review.

**Review by jmjatlanta — APPROVED** [review]

**Comment by jmjatlanta** [comment]
There may be an edge case here that we may want to fix, but the chances of it happening are very remote. Scenario: Destructor called, then rotate_files gets the lock.

I am okay with that being there, as the file_appender is typically only destroyed at program exit, and the problem has probably been there from before this issue was opened.

**Comment by cogutvalera** [comment]
@pmconrad @jmjatlanta Thanks guys !

**Comment by cogutvalera** [comment]
Thanks !
